### ISSUE TITLE
closing blocked IO across threads hangs (and Ruby-only workaround is dangerous)

### ISSUE THREAD

**Issue by ghost** [rubinius/rubinius#2757]
```
a, b = IO.pipe

# start background thread to block on read,
# on MRI, this will register the thread as being blocked on a
# (th->waiting_fd if you read MRI source)
th = Thread.new { a.read }

# wait for the thread to get blocked on a.read
sleep 1

# What MRI does may not be easily doable in rbx due to lack of GVL,
# but I'm documenting here for reference:
#
# MRI tracks all threads blocked in IO via rb_thread_io_blocking_region.
# This allows the IO#close implementation to walk through the thread
# list and send a signal to all threads waiting on the to-be-closed
# IO object.
#
# MRI does the following on close:
#   1) invalidate IO object (so further operations will IOError)
#   2) walk thread list and signal any threads blocked on closing IO,
#      threads will retry IO operation on invalid IO object
#   3) finally issue the close() syscall
a.close

# As a Ruby programmer, I would not mind issuing th.run manually
# after issuing a.close above.
#
# Unfortunately, uncommenting th.run causes EBADF on th.join.
# An EBADF on read() is extremely dangerous in a multithreaded program.
# rbx should revalidate the associated FD of an IO object on any
# thread that is hit by EINTR.

# th.run

th.join # raises IOError on MRI, but hangs in rbx

# Basically, I would not mind having to issue my own th.run call
# after a.close if MRI compatibility cannot be 100%, but the current
# EBADF situation is extremely dangerous in a multithreaded program.
#
# Also, note I prefer blocking I/O in some situations since the kernel
# can make better scheduling decisions when multiple threads sleep on the
# same IO operation than if I used IO.select, so please don't add
# select/poll/ppoll before syscalls (on Linux).
```
